Fix #6449: Added check for CODEOWNERS to ensure that the most important rules are at the end.

YashJipkate opened the pull request:
<!--
  - Thanks for submitting code to Oppia! Please fill out the following as part of
  - your pull request so we can review your code more easily.
  -->

## Explanation
<!--
  - Explain what your PR does. If this PR fixes an existing bug, please include
  - "Fixes #bugnum:" in the explanation so that GitHub can auto-close the issue
  - when this PR is merged.
  -->
Fixes #6449. A check is added to ensure the most important rules/patterns are at the bottom of the CODEOWNERS file. 

## Checklist
- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The linter/Karma presubmit checks have passed.
  - These should run automatically, but if not, you can manually trigger them locally using `python scripts/pre_commit_linter.py` and `bash scripts/run_frontend_tests.sh`.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR follows the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide).
- [x] The PR is assigned to an appropriate reviewer.
  - If you're a new contributor, please ask on [Gitter](https://gitter.im/oppia/oppia-chat) for someone to assign a reviewer.
  - If you're not sure who the appropriate reviewer is, please assign to the issue's "owner" -- see the "talk-to" label on the issue.

YashJipkate committed:
Added imp rules check

YashJipkate committed:
Removed unnecessary comment

YashJipkate commented:
Hi, @apb7 Could you PTAL!
Thanks!

YashJipkate commented:
Hi @apb7 Ping!

apb7 commented on scripts/pre_commit_linter.py:
```diff
@@ -268,6 +268,14 @@
 CODEOWNER_FILE_PATHS = ['./app.yaml', './manifest.json']
 
+CODEOWNER_IMPORTANT_PATHS = [
+    '/manifest.json',
```
Let's alphabetize this list.

YashJipkate replied:
Done!

apb7 commented on scripts/pre_commit_linter.py:
```diff
@@ -2178,6 +2190,23 @@ def _check_codeowner_file(self):
 
+            # Checks that the most important rules/patterns are at the bottom
+            # of the CODEOWNERS file.
+            important_path_match_bool_list = ([
```
A couple of points here:
1. Please add comments explaining why we are populating this list.
2. Where are we checking whether the paths in `CODEOWNER_IMPORTANT_PATHS` are towards the end of the CODEOWNERS file?

YashJipkate replied:
Added the comments. PTAL!

apb7 replied:
I think you haven't answered this:
>Where are we checking whether the paths in CODEOWNER_IMPORTANT_PATHS are towards the end of the CODEOWNERS file?

You are only checking the lines in CODEOWNER_IMPORTANT_PATHS as far as I understand.

YashJipkate replied:
Yes, we are checking for the ones in `CODEOWNER_IMPORTANT_PATHS` which contains the important rules.
Does this seem wrong?

apb7 replied:
Thanks for the explanation. Looks good!

apb7 commented on scripts/pre_commit_linter.py:
```diff
@@ -2178,6 +2190,23 @@ def _check_codeowner_file(self):
+                for index, bool_value in enumerate(
+                        important_path_match_bool_list):
+                    if not bool_value:
+                        print ('%s --> Please ensure that pattern \'%s\' is'
```
Unclear error message. Please re-frame to something like:
`Please ensure that the rule <pattern> lies towards the bottom of the CODEOWNERS file here <codeowner_filepath> since it is an important rule.`

YashJipkate replied:
Done!

apb7 reviewed (changes requested):
Hi @YashJipkate, sorry for the delay. Please manually check these changes by deliberately making such an error, locally. Thanks!

YashJipkate committed:
Added comments, alphabetize list and improved error message

apb7 commented:
@YashJipkate, did you do this as well?
> Please manually check these changes by deliberately making such an error, locally. Thanks!

Please answer all queries before requesting another review :)

YashJipkate commented:
Sorry for that @apb7.
Yes I have tried this manually on my local by moving the important rules upward such that they are not at the end of the file, and the linter reported the same too.
Thanks!

apb7 reviewed (approved):
LGTM, thanks, @YashJipkate!

YashJipkate commented:
Hi @apb7, Is this good to merge?
Thanks!

DubeySandeep commented on scripts/pre_commit_linter.py:
```diff
@@ -268,6 +268,14 @@
 
 CODEOWNER_FILE_PATHS = ['./app.yaml', './manifest.json']
 
+CODEOWNER_IMPORTANT_PATHS = [
```
Can you please add a comment here and also in the CODEOWNERS file saying that this need to be in sync?

YashJipkate replied:
Yeah sure. Will do that!

DubeySandeep commented on scripts/pre_commit_linter.py:
```diff
@@ -2178,6 +2192,30 @@ def _check_codeowner_file(self):
                             print ('WARNING! %s/%s is not covered under '
                                    'CODEOWNERS' % (root, file_name))
 
+            # Checks that the most important rules/patterns are at the bottom
```
I think to check code owner function is already 150~ lines big can we create a sperate function for checking important rule or pattern and call this over here? will it be possible?   

This will also move this many comments into the new function docstring!

YashJipkate replied:
Yeah sounds good. Will do that!
Also, @apb7, do you agree on this too? (since you are always in favor of clubbing similar functionalities to a pre-existing function if any)

apb7 replied:
I agree with @DubeySandeep here. We can modularize this code. (As an FYI -- I am in favor of clubbing similar checks so that we do not end up repeating stuff. If there's no repetition, then it's totally fine :))

DubeySandeep replied:
@YashJipkate, is this comment resolved?

YashJipkate replied:
Yes @DubeySandeep This is resolved now. Sorry for the delay.
PTAL! Thanks

DubeySandeep commented on scripts/pre_commit_linter.py:
```diff
@@ -2178,6 +2192,30 @@ def _check_codeowner_file(self):
+            # matching the last lines of the CODEOWNERS file with the list of
+            # the important CODEOWNERS paths.
+            important_path_match_bool_list = ([
+                imp_path in path_lines[-1 * len(
```
Is someone adds a new pattern(important) in the top of the list, will this check ask users to add this list over here or will this check fail?

YashJipkate replied:
No, the check will fail. The 'n'-list and the last 'n' lines need to be in sync.

DubeySandeep replied:
![image](https://user-images.githubusercontent.com/16653571/56900733-293c8380-6ab4-11e9-94c1-2c4535b3e3bf.png)
I tried adding a new file in the CODEOWNERS critical section. (without doing any changes in pre_commit_linter.py)

What should be the expected behavior?
I got this:
![image](https://user-images.githubusercontent.com/16653571/56900805-5426d780-6ab4-11e9-986b-4ff7e45bf37d.png)

YashJipkate replied:
This is happening because of the new important rules that are not in sync with the `CODEOWNER_IMPORTANT_PATHS` list. The (outdated) list checks for the last n-lines of the CODEOWNERS file in which the erring pattern is not found and hence the error.

DubeySandeep replied:
I'm happy that it's failing but the error message is something unrelated. Can we do something for that?

YashJipkate replied:
I have modified the error message. PTAL!
Thanks.

DubeySandeep commented on scripts/pre_commit_linter.py:
```diff
@@ -2178,6 +2192,30 @@ def _check_codeowner_file(self):
+            # This condition checks that all the values in the boolean list are
+            # True. This ensures that the last 'n' lines of the CODEOWNERS file
+            # matches with the 'n'-element list of important CODEOWNER paths.
+            if not all(important_path_match_bool_list):
```
This will also look for all the value in the list i.e, what we have done in line 2209, if yes then can we just have the for loop? (is there any benefit of doing this check?)

YashJipkate replied:
I added this check so that the loop runs only when it is needed. At present the list is small so it may not seem that big a problem but in future when it enlarges it might be redundant to unnecessary run the loop.
What do you think?

DubeySandeep replied:
In any case, we are going to run through all the values, isn't it?

Like if there is any false in the list we will cover the entire list through the for loop and if all the values are true we are going to loop through all the values via `all(...)`, right?

YashJipkate replied:
Yeah, you are right about this. Will remove this. Thanks for pointing out!

DubeySandeep reviewed (commented):
@YashJipkate, I went through this PR and left a few comments can you please check! :)

YashJipkate committed:
Merge branch 'develop' into codeowner-imp-rules

YashJipkate committed:
Review addressed

DubeySandeep reviewed (changes requested):
(I'm marking this to "Request changes")

seanlip reviewed (approved):
LGTM (but only as code owner of CODEOWNERS :) )

YashJipkate committed:
Improved error messages

DubeySandeep commented:
Hi @YashJipkate, Is this PR ready for review?

YashJipkate commented:
Hi @DubeySandeep, yes all the changes are made and I am waiting for @apb7 here(https://github.com/oppia/oppia/pull/6560#discussion_r279369339)

apb7 commented:
Hi @YashJipkate, I have already addressed this https://github.com/oppia/oppia/pull/6560#discussion_r279369339. @DubeySandeep, can you please approve and merge this? Thanks!

YashJipkate committed:
Merge and break codeowners check function

DubeySandeep commented:
@YashJipkate, Lint test is failing PTAL!

DubeySandeep commented on scripts/pre_commit_linter.py:
```diff
@@ -2147,42 +2166,103 @@ def _check_for_copyright_notice(self):
             summary_messages = []
             # Checks whether every pattern in the CODEOWNERS file matches at
             # least one dir/file.
+            path_patterns = []
```
is this variable used anywhere?

YashJipkate replied:
No, removed!

DubeySandeep commented on scripts/pre_commit_linter.py:
```diff
@@ -2234,11 +2314,14 @@ def _check_codeowner_file(self):
                         failed = True
 
+            failed = self.check_for_important_rules_at_bottom_of_codeowners(
+                imp_rules_in_critical_section, failed)
```
Is that `failed` parameter is used in `check_for_important_rules_at_bottom_of_codeowners` in anyway? If not, then can we do AND operation between the existing `failed` result and the return value of `check_for_important_rules_at_bottom_of_codeowners`?

YashJipkate replied:
The `failed` parameter is the same as in the lines above. There is no need to perform an AND operation because the `failed` parameter is always set to True whenever its value is changed and never set to False. So if  `failed` evaluates to True due to earlier checks, it will still be True since it is never set to False.

seanlip replied:
See comment above.

DubeySandeep commented on scripts/pre_commit_linter.py:
```diff
@@ -2234,11 +2314,14 @@ def _check_codeowner_file(self):
                     _MESSAGE_TYPE_FAILED)
             else:
-                summary_message = '%s  CODEOWNERS file check passed' % (
+                summary_message = '%s   CODEOWNERS file check passed' % (
```
I think this is not an expected change, right?

YashJipkate replied:
I did this because of the inconsistency with the whitespaces in messages like this. But I think I would revert this if it doesn't seem right.

DubeySandeep commented on scripts/pre_commit_linter.py:
```diff
@@ -2147,42 +2166,103 @@ def _check_for_copyright_notice(self):
+        of the CODEOWNERS file.
+
+        Arguments:
+            imp_rules_in_critical_section: list(str). List of the important
```
Can we rename this parameter to `imp_patterns` or `imp_paths` (something around this to make it generic)?

YashJipkate replied:
Done.

DubeySandeep commented on scripts/pre_commit_linter.py:
```diff
@@ -2147,42 +2166,103 @@ def _check_for_copyright_notice(self):
+        Arguments:
+            imp_rules_in_critical_section: list(str). List of the important
+                paths/rules for CODEOWNERS file.
+            failed: bool. Current status of failure.
```
Ditto, do we need this parameter?

YashJipkate replied:
Yes we do need this parameter according to the approach I followed.

seanlip replied:
This is a strange API design. I don't think you should have it. The failure status should be returned from this function independently and the caller should be responsible for determining the final status.

YashJipkate replied:
Yes, I have done it now!

DubeySandeep commented on scripts/pre_commit_linter.py:
```diff
@@ -2147,42 +2166,103 @@ def _check_for_copyright_notice(self):
+                imp_path in CODEOWNER_IMPORTANT_PATHS
+                for imp_path in imp_rules_in_critical_section])
+            for index, bool_value in enumerate(
+                    important_path_match_bool_list):
```
Can we traverse the `imp_rules_in_critical_section` and do an if check in the loop? (Instead of creating a new list and lopping over that new list) I think having just one loop will make it readable and clear. Let me know if there's an issue with that!

YashJipkate replied:
But I think that having a separate variable for this would be more readable similar to the modularization we did for this function as we concluded that splitting up would be better for the end user.

YashJipkate committed:
Fix lint

DubeySandeep commented on scripts/pre_commit_linter.py:
```diff
@@ -2147,42 +2166,103 @@ def _check_for_copyright_notice(self):
+                    failed = True
+        else:
+            important_path_match_bool_list = ([
+                imp_path in imp_rules_in_critical_section
```
Ditto.

DubeySandeep commented on scripts/pre_commit_linter.py:
```diff
@@ -2147,42 +2166,103 @@ def _check_for_copyright_notice(self):
+            for index, bool_value in enumerate(
+                    important_path_match_bool_list):
+                if not bool_value:
+                    print ('%s --> Please ensure that the rule \'%s\' lies '
```
Do we have to print this through _redirect_stdout? (not sure!)

YashJipkate replied:
No I dont think so since all the print statements are printed without _redirect_stdout.

DubeySandeep reviewed (changes requested):
@YashJipkate, I took a pass and left a few comments PTAL!

I was thinking to check the diff between feature-branch and upstream/develop for the .github/CODEOWNERS file and validate the difference in the critical section, instead of maintaining a consistent list of critical patterns in pre_commit_linter.py But I think this will make the pre_commit_linter.py dependent on an internet connection and also not sure whether this will work on CI.

@apb7, can we follow the above-mentioned pattern for the check?

Note: I'm happy with the proposed structure, I suggest the above structure just to avoid maintain a list in the pre_commit_linter.py.

YashJipkate committed:
Merge develop

YashJipkate committed:
Address review

seanlip commented on scripts/pre_commit_linter.py:
```diff
@@ -2215,42 +2234,102 @@ def _check_bad_patterns(self):
         return summary_messages
 
+    def check_for_important_rules_at_bottom_of_codeowners(
+            self, imp_patterns, failed):
```
imp_patterns --> important_patterns. Don't use abbreviations.

Ditto elsewhere, e.g. imp_path, imp_rules, etc.

YashJipkate replied:
Done.

seanlip commented on scripts/pre_commit_linter.py:
```diff
@@ -2215,42 +2234,102 @@ def _check_bad_patterns(self):
+            bool. Status of failure.
+        """
+
+        # The following condition checks for extra rules in the critical
```
Would it perhaps be cleaner to:

- Make a set A of important paths from the constant in this file
- Make a set B of important paths from the CODEOWNERS file
- Use set difference operations (A-B, B-A) in Python to figure out what's missing in either

I think that would be cleaner and easier to read than printing everything individually and doing lots of for loops.

DubeySandeep replied:
@seanlip I was too thinking of doing this cleanly using `set`, but the only issue over here is with repeated patterns, they won't get caught! (If we want to catch repeated pattern we can create `check_no_duplicate iteam_in_list` before creating a `set`.)

(Repeated pattern in the ciriticle path patterns.)

seanlip replied:
Good call! A common pattern for this is to start by doing it with a list, then convert it to set and check if the lengths match.

YashJipkate replied:
Done. PTAL if it matches with your thoughts @seanlip @DubeySandeep

seanlip reviewed (changes requested):
HI @YashJipkate, took a look and left some comments, PTAL. Thanks!

YashJipkate committed:
Address review

apb7 commented:
Hi @YashJipkate, have you addressed the last round of review comments? If yes, can you please reply to them inline so that the reviewers have an idea. This PR has been standing for quite some time now. Can you please look into this? Thanks!

YashJipkate commented:
Hi @apb7, Sorry for this PR getting stale. I have addressed all the comments except this https://github.com/oppia/oppia/pull/6560#discussion_r288764451 since that requires complete restructure of the method I have used.

seanlip commented:
Hi @YashJipkate, I think you can still address that comment. The change is not that big. Are you planning to do so?

YashJipkate committed:
Merge branch 'develop' into codeowner-imp-rules

YashJipkate committed:
Changed working rule to use set operations

YashJipkate committed:
Fix lint

YashJipkate commented:
Hi @seanlip Sorry for the delay, I have changed the working rule to make way for sets. PTAL!
Thanks!

seanlip commented on scripts/pre_commit_linter.py:
```diff
@@ -2188,42 +2207,99 @@ def _check_bad_patterns(self):
+
+        Arguments:
+            important_patterns: list(str). List of the important
+                paths/rules for CODEOWNERS file.
```
Here you say "paths/rules", but in the docstring you say "rules/patterns", and to make things even more confusing, this arg is called "important_patterns".

Please could you figure out what is actually being referenced, and amend the docstrings/argname/descriptions accordingly?

YashJipkate replied:
I have generalized "patterns".

seanlip commented on scripts/pre_commit_linter.py:
```diff
@@ -2188,42 +2207,99 @@ def _check_bad_patterns(self):
+                paths/rules for CODEOWNERS file.
+
+        Returns:
+            bool. Status of failure.
```
bool. Whether the CODEOWNERS "important rule" check passes.

YashJipkate replied:
"- check fails" would be more appropriate I guess since the bool actually has the status of failure.

seanlip commented on scripts/pre_commit_linter.py:
```diff
@@ -2188,42 +2207,99 @@ def _check_bad_patterns(self):
+                  ' section.' % CODEOWNER_FILEPATH)
+            failed = True
+        if len(codeowner_important_paths_set) != len(CODEOWNER_IMPORTANT_PATHS):
+            print('scripts/pre_commit_linter.py --> Duplicate pattern found in '
```
Nit: pattern --> patterns for consistency with previous message?

YashJipkate replied:
Done!

seanlip commented on scripts/pre_commit_linter.py:
```diff
@@ -2188,42 +2207,99 @@ def _check_bad_patterns(self):
+            print('scripts/pre_commit_linter.py --> Duplicate pattern found in '
+                  'CODEOWNER_IMPORTANT_PATHS list.')
+            failed = True
+        # Check missing rules by set difference operation.
```
Put a blank line above this. Divide your code into meaningful "blocks".

YashJipkate replied:
Done!

seanlip commented on scripts/pre_commit_linter.py:
```diff
@@ -2188,42 +2207,99 @@ def _check_bad_patterns(self):
+            failed = True
+        for rule in list_minus_critical_rule_section_set:
+            print('%s --> Rule \'%s\' is not present in the \'Critical files\' '
+                  'section. Please place it under the \'Critical files\' since '
```
since --> section since

YashJipkate replied:
Done!

seanlip reviewed (changes requested):
This looks good to me, just some very minor comments left. Thanks @YashJipkate!

YashJipkate committed:
Address review

seanlip reviewed (approved):
LGTM. Thanks @YashJipkate !

seanlip commented:
@DubeySandeep PTAL. Thanks!

YashJipkate committed:
Fix lint

seanlip commented:
@YashJipkate qq -- don't the lint checks run before you push? Just wondering how that slipped through and how we can prevent it in the future.

/cc @oppia/dev-workflow-team

YashJipkate commented:
Actually, in this copy of my repo, there was some problem regarding the python version (since this my new system) and hence the linter does not work (probably). The other copy works fine as it had the correct python version installed.

seanlip commented:
Ah, ok. If you feel like we need to fix the pre-commit linter to guard against this issue, it might be worth doing that. Might save a lot of time in future.

YashJipkate commented:
Thanks @seanlip. But fortunately, I no longer use that copy and use the one which has the linter running. This is an old PR and hence was in the old copy.

seanlip commented:
OK cool! Sounds good then :D

YashJipkate commented:
@DubeySandeep The checks pass now. Could this be merged?
Thanks!
